### PR TITLE
[DTensor] Fix non-contiguous local tensor after Shard->Replicate redistribute

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #174134

When redistributing from `Shard(dim)` to `Replicate()` with uneven shard
sizes (dimension size not divisible by world_size), the local tensor
becomes non-contiguous. This happens because `_maybe_unpad_tensor` uses
`tensor.narrow()` which returns a view with incorrect strides.

For example, redistributing a tensor of shape [4, 31, 10] from Shard(1)
to Replicate results in stride (320, 10, 1) instead of the expected
contiguous stride (310, 10, 1).

The fix adds a contiguity check after unpadding, matching the existing
pattern in `_maybe_pad_tensor` which already ensures contiguity.

Fixes #173041

Claude

### CODE REVIEW

**Comment by fmassa** [comment]
Discussed on chat, while I'm ok always returning contiguous tensors, I wonder if we shouldn't instead fix the operators in PyTorch that don't handle non-contiguous tensors?

**Comment by fmassa** [comment]
Basically, I thought that we had the following contract: all ops should support arbitrarily-strided tensors, even if this means calling `.contiguous()` first thing in the op.

For the error the user reported, I believe he was facing it because `nn.Linear` desugars into `view -> mm -> view`, and `view` doesn't support non-contiguous tensors. I wonder if the simplest fix here would be to change it to call `.reshape` instead

**Review by zpcore — APPROVED** [review]
